File: 198.house-robber.py
# @leet start
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) <= 2:
            return max(nums)

        memo = [0 for _ in nums]
        for i in range(len(nums)):
            one, two = 0, 0
            if i - 2 >= 0:
                one = memo[i - 2]
            if i - 3 >= 0:
                two = memo[i - 3]

            memo[i] = nums[i] + max(one, two)

        return max(memo[-1], memo[-2])


# @leet end

File: 198.house-robber.ts
// @leet start
function rob(nums: number[]): number {
	let l = nums.length;
	if (nums.length < 2) {
		return nums[0];
	}

	let mem = [nums[0], nums[1]];

	if (2 < l) {
		mem.push(nums[0] + nums[2]);
	}

	for (let i = 3; i < l; i++) {
		mem.push(Math.max(mem[i - 2], mem[i - 3]) + nums[i]);
	}

	return Math.max(mem[l-1], mem[l-2]);
    
};
// @leet end
